plutus-scb/src/Cardano/Protocol/Socket/Server.hs

```diff
@@ -0,0 +1,215 @@
+-- be shared with the thread that starts it.
+-- The thread that starts the server node will also handle the controller
+-- effects for the node (adding blocks).
+startServerNode :: FilePath -> MVar AppState -> IO Void
+startServerNode sockPath stateMV = withIOManager $ \iocp -> do
+    networkState <- newNetworkMutableState
+    _ <- async $ cleanNetworkMutableState networkState
+    withServerNode
+      (localSnocket iocp "")
+      nullNetworkServerTracers
+      networkState
+      (AcceptedConnectionsLimit maxBound maxBound 0)
+      (localAddressFromPath sockPath)
+      unversionedHandshakeCodec
+      cborTermVersionDataCodec
+      (\(DictVersion _) -> acceptableVersion)
+      (simpleSingletonVersions UnversionedProtocol
+                               (NodeToNodeVersionData $ NetworkMagic 0)
+                               (DictVersion nodeToNodeCodecCBORTerm)
+                               (\_peerid -> SomeResponderApplication app))
+      nullErrorPolicies
+      $ \_ serverAsync -> wait serverAsync
+
+    where
+      -- | Applications are where the protocol itself is described. This is an
+      -- application with 2 multiplexed protocols (ChainSync and TxSubmission,
+      -- with BlockFetch coming soon).
+      app :: OuroborosApplication 'ResponderApp
+                                  LBS.ByteString
+                                  IO Void ()
+      app = nodeApplication chainSync txSubmission
+
+      -- | This is boilerplate that wires in a protocol using the `chainSyncServer`
+      -- function for the heavy lifting.
+      chainSync :: RunMiniProtocol 'ResponderApp LBS.ByteString IO Void ()
+      chainSync =
+          ResponderProtocolOnly $
+          MuxPeer
+            (contramap show stdoutTracer)
+            codecChainSync
+            (ChainSync.chainSyncServerPeer (chainSyncServer stateMV))
+
+      -- | This is boilerplate that wires in a protocol using the `txSubmissionServer`
+      -- function for the heavy lifting.
+      txSubmission :: RunMiniProtocol 'ResponderApp LBS.ByteString IO Void ()
+      txSubmission =
+          ResponderProtocolOnly $
+          MuxPeer
+            (contramap show stdoutTracer)
+            codecTxSubmission
+            (TxSubmission.localTxSubmissionServerPeer (pure $ txSubmissionServer stateMV))
+
+-- | This is the local submission protocol, server side. It keeps on receiving new transactions
+-- via the node protocol, and adds them to the transaction pool.
+txSubmissionServer :: MVar AppState
+                   -> TxSubmission.LocalTxSubmissionServer Tx String IO ()
+txSubmissionServer state = idleState
+    where
+      idleState :: TxSubmission.LocalTxSubmissionServer Tx String IO ()
+      idleState =
+        TxSubmission.LocalTxSubmissionServer {
+          TxSubmission.recvMsgSubmitTx =
+            \tx -> do
+                st <- takeMVar state
+                putMVar state (over (chainState . txPool) (tx :) st)
```

Comment:
`modifyMVar`? I think that's got better atomicity (or, well, maybe the same, but it's clearer)

Reply:
Nit: let's put a `addTxToPool :: Tx -> TxPool -> TxPool` function to `Chain`. The current implementation is currently just `:`, but I think it makes stuff like this much easier to read.

Reply:
The change is getting fairly big. Would it be okay if I create a new issue and deal with it in the next sprint?

Reply:
Sure, up to you.

Reply:
Added issue https://github.com/input-output-hk/plutus/issues/2140 to track the `addTxToPool` function.